compiler/rustc_expand/src/mbe/transcribe.rs

```diff
@@ -763,19 +764,29 @@ fn extract_ident<'a>(
             }
             return Ok(nt_ident.to_string());
         }
-        if let ParseNtResult::Tt(TokenTree::Token(
-            Token { kind: TokenKind::Ident(token_ident, is_raw), .. },
-            _,
-        )) = pnr
-        {
-            if let IdentIsRaw::Yes = is_raw {
-                return Err(dcx.struct_span_err(ident.span, RAW_IDENT_ERR));
+
+        if let ParseNtResult::Tt(TokenTree::Token(Token { kind, .. }, _)) = pnr {
+            if let TokenKind::Ident(symbol, is_raw) = kind {
+                if let IdentIsRaw::Yes = is_raw {
+                    return Err(dcx.struct_span_err(ident.span, RAW_IDENT_ERR));
+                }
+                return Ok(symbol.to_string());
             }
-            return Ok(token_ident.to_string());
+
+            if let TokenKind::Literal(Lit { kind: LitKind::Str, symbol, suffix: None }) = kind {
+                return Ok(symbol.to_string());
+            }
+        }
+
+        if let ParseNtResult::Nt(nt) = pnr
+            && let Nonterminal::NtLiteral(expr) = &**nt
+            && let ExprKind::Lit(Lit { kind: LitKind::Str, symbol, suffix: None }) = &expr.kind
+        {
+            return Ok(symbol.to_string());
         }
     }
     Err(dcx.struct_span_err(
         ident.span,
-        "`${concat(..)}` currently only accepts identifiers or meta-variables as parameters",
+        "`${concat(..)}` currently only accepts identifiers, token trees or literals",
```

Comment:
```suggestion
        "`${concat(..)}` currently only accepts identifiers or string literals",
```
I think we already have a similar message somewhere which can be reused.

Reply:
This error is more specialized because it is only dealing with meta-variables.

The message has been changed to a more intuitive description.